### PR TITLE
feat: End conversation from widget

### LINKED ISSUE #2828: End conversation from chat widget
**Context:**

Provide an option to the user to end the conversation once they are done with it. This would move the conversation from the queue.

**Changes:**
- [ ] Add a dropdown on the widget to select the end conversation button. Design attached. Move the existing open in a new window button to the dropdown.
- [ ] Add an API to resolve the conversation.
- [ ] Add an activity message "Conversation was resolved by %{user_name}".

**Design:**

end-chat(2)

### PR DESCRIPTION
**Pull request by aswindevps**
## Description

This change adds a dropdown on the widget. So that the end-users can resolve the conversation from their side. 

Fixes #2828

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## How Has This Been Tested?

https://user-images.githubusercontent.com/5848740/147570029-b772c02f-e030-4d6a-9956-dc0e6ecc9624.mp4

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/javascript/widget/components/dropdown/DropdownMenu.vue [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    },
+  },
+  mounted() {
+    const onEscape = e => {
+      if (e.key === 'Esc' || e.key === 'Escape') {
+        this.open = false;
+      }
+    };
+
+    document.addEventListener('keydown', onEscape);
```
Comment (fayazara) [inline app/javascript/widget/components/dropdown/DropdownMenu.vue:59]: I think we should add a [removeEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener) too on the [beforeDestroy](https://vuejs.org/v2/api/#beforeDestroy) hook.

**Review by fayazara — CHANGES_REQUESTED** [review]
Have added a small comment, otherwise looks good. Will also do a quick test and update.

File: config/locales/en.yml [hunk 2]
```diff
@@ -60,60 +60,61 @@ en:
...
       description: "Webhook events provide you the realtime information about what's happening in your account. You can make use of the webhooks to communicate the events to your favourite apps like Slack or Github. Click on Configure to set up your webhooks."
     dialogflow:
-      name: "Dialogflow"
-      description: "Build chatbots using Dialogflow and connect them to your inbox quickly. Let the bots handle the queries before handing them off to a customer service agent."
+      name: 'Dialogflow'
+      description: 'Build chatbots using Dialogflow and connect them to your inbox quickly. Let the bots handle the queries before handing them off to a customer service agent.'
     fullcontact:
-      name: "Fullcontact"
-      description: "FullContact integration helps to enrich visitor profiles. Identify the users as soon as they share their email address and offer them tailored customer service. Connect your FullContact to your account by sharing the FullContact API Key."
+      name: 'Fullcontact'
```
Comment (pranavrajs) [inline config/locales/en.yml:119]: @aswindevps Can you please revert the changes which are not related to your PR?

File: app/javascript/widget/components/dropdown/DropdownMenu.vue (hunk 1 above)
Comment (pranavrajs) [reply app/javascript/widget/components/dropdown/DropdownMenu.vue:59]: yeah it is required

File: app/javascript/widget/components/dropdown/DropdownMenu.vue [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  },
+};
+</script>
+<style scoped lang="scss">
+@import '~widget/assets/scss/variables.scss';
+
+.menu-content {
+  width: max-content;
+  border: 1.5px solid $color-border;
```
Comment (pranavrajs) [inline app/javascript/widget/components/dropdown/DropdownMenu.vue:68]: I'm not sure whether we should go with 1.5? Any particular reason for the same? Can't we tailwind classes instead of this.

File: app/javascript/widget/components/dropdown/DropdownMenuItem.vue [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      default: () => {},
+    },
+  },
+};
+</script>
+<style scoped lang="scss">
+@import '~widget/assets/scss/variables.scss';
+
+.menu-item {
+  border-bottom: 1.5px solid $color-border;
```
Comment (pranavrajs) [inline app/javascript/widget/components/dropdown/DropdownMenuItem.vue:57]: 1.5px seems off. Any particular reason to choose the measurement?

File: app/javascript/widget/components/HeaderActions.vue [hunk 5]
```diff
@@ -42,6 +71,15 @@ export default {
     showHeaderActions() {
       return this.isIframe || this.isRNWebView;
     },
+    isResolved() {
+      if (this.conversationAttributes.status) {
```
Comment (pranavrajs) [inline app/javascript/widget/components/HeaderActions.vue:75]: Do we need to check this? It would available always right?

**Review by pranavrajs — COMMENTED** [review]
Nice work ❤️. Please see the individual comments.

**Comment by muhsin-k** [comment]
@aswindevps The "End conversation" showing only when I refresh the widget
https://user-images.githubusercontent.com/12408980/147644623-3b1f2225-97de-4f10-aca1-391ba7dc7fed.mov

**Comment by nithindavid** [comment]
@aswindevps The delete icon seems misleading for resolve, Can we use Tick icon just like on dashboard side ?

File: app/javascript/widget/components/dropdown/DropdownMenu.vue (hunk 3 above)
Comment (aswindevps) [reply app/javascript/widget/components/dropdown/DropdownMenu.vue]: Updated to tailwind classes

File: app/javascript/widget/components/dropdown/DropdownMenuItem.vue (hunk 4 above)
Comment (aswindevps) [reply app/javascript/widget/components/dropdown/DropdownMenuItem.vue]: Updated to tailwind classes

File: app/javascript/widget/components/HeaderActions.vue (hunk 5 above)
Comment (aswindevps) [reply app/javascript/widget/components/HeaderActions.vue]: Updated the logic.

**Comment by aswindevps** [comment]
**Testing scenarios**

1. Start and resolve conversation from the end-user side.
2. Start a conversation with the end-user and resolve from the Agent side.
3. Initiate a conversation using the campaign. 
    a) Resolve from the end-user side.
    b) Resolve from the agent side.

Test the above scenarios with CSAT and Allow messages after conversation resolved flags.

Please add more scenarios if required.

File: app/javascript/widget/components/dropdown/DropdownMenu.vue [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    },
+  },
+  data() {
+    return {
+      isOpen: false,
+    };
+  },
+  watch: {
+    // eslint-disable-next-line func-names
+    open: function() {
```
Comment (nithindavid) [inline app/javascript/widget/components/dropdown/DropdownMenu.vue:59]:
```suggestion
    open() {
```

**Review by nithindavid — CHANGES_REQUESTED** [review]
please see the minor change.

File: app/javascript/widget/components/dropdown/DropdownMenu.vue (hunk 6 above)
Comment (nithindavid) [reply app/javascript/widget/components/dropdown/DropdownMenu.vue:59]: // eslint-disable-next-line func-names Would no longer be needed.

**Review by sojan-official — APPROVED** [review]
Backend code LGTM

**Review by tejaswinichile — DISMISSED** [review]
Ruby changes LGTM 💯

**Comment by pranavrajs** [comment]
@aswindevps Can you please take a lead and figure out what needs to be done here?
